### PR TITLE
Support Delayed Deletion for Vmwareengine Private Cloud

### PR DESCRIPTION
**Pull request by swamitagupta**
1. Adds support to delay the Private Cloud deletion . More information about the intended behavior can be found [here](https://cloud.google.com/vmware-engine/docs/private-clouds/howto-delete-private-cloud#delete_a_private_cloud). 
2. Modifies the region for vmwareengine resource tests to region 'me-west1' with allocated quota.
3. Facilitates type change from `TIME_LIMITED` to `STANDARD` for multi-node Private Clouds

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
vmwareengine: added `deletion_delay_hours` field to `google_vmwareengine_private_cloud` resource
```
```release-note:enhancement
vmwareengine: support type change from `TIME_LIMITED` to `STANDARD` for multi-node `google_vmwareengine_private_cloud` resource
```

### CODE REVIEW

**Comment by swamitagupta** [comment]
This PR intends to solve #10764

**Comment by swamitagupta** [comment]
Hi @melinath, 
Wanted to let you know that I am in the process of testing various CUJs locally and will update you once the changes have successfully passed the tests and are ready to be reviewed for submission.

**Review by melinath — COMMENTED** [review]
ah okay - in that case, marking as reviewed for now.

**Review by melinath — COMMENTED** [review]
It looks like this might not be ready for review yet; re-request review if you do want me to take a look.

File: mmv1/templates/terraform/post_delete/private_cloud.go.erb [hunk 1]
```diff
@@ -26,9 +26,19 @@ privateCloudPollRead := func(d *schema.ResourceData, meta interface{}) transport
                 RawURL: url,
                 UserAgent: userAgent,
             })
+            // if resource does not exist
             if err != nil {
+                log.Printf("[DEBUG] The Private cloud has been successfully deleted.")
```
Comment (melinath) [inline mmv1/templates/terraform/post_delete/private_cloud.go.erb:31]: this if statement actually indicates there was an error during deletion.

File: mmv1/templates/terraform/post_delete/private_cloud.go.erb [hunk 2]
```diff
@@ -26,9 +26,19 @@ privateCloudPollRead := func(d *schema.ResourceData, meta interface{}) transport
...
                 return res, err
             }
+            // if resource exists but is marked for deletion
+            log.Printf("[DEBUG] Fetching state of the private cloud.")
+            v, ok := res["state"]
+            if ok && v.(string) == "DELETED" {
+                log.Printf("[DEBUG] The Private cloud has been successfully marked for delayed deletion.")
+                return res, errors.New("PrivateCloud is marked for deletion")
```
Comment (melinath) [inline mmv1/templates/terraform/post_delete/private_cloud.go.erb:39]: I think you might want to return nil, nil here to signify that the resource was successfully put in a deleted state.

**Review by melinath — CHANGES_REQUESTED** [review]

**Comment by melinath** [comment]
Put down some initial thoughts, but I'm handing off the review since it looks like I won't be able to wrap it up before I'm OOO for a bit.

File: mmv1/products/vmwareengine/PrivateCloud.yaml [hunk 3]
```diff
@@ -89,6 +91,17 @@ parameters:
     url_param_only: true
     description: |
       The ID of the PrivateCloud.
+  - !ruby/object:Api::Type::NestedObject
+    name: 'deletionParameters'
```
[the remark this answers is not in the document]
Comment (swamitagupta) [reply mmv1/products/vmwareengine/PrivateCloud.yaml:95]: I initially planned to add only the `deletion_delay_hours` parameter. However, this made me unable to achieve the following behavior:
- if user enters no value, pass the delete API url without the delay hr mask. This would add the default delay hour which is set at the api level. We dont want to hardcode this value at the terraform level so that it can be easily modified and is at one place. 
- if the user adds any value (0 and non zero), use this value for the delay hr mask in the delete url.  

Since Terraform cant distinguish between empty and zero values, I created a nested object so that I can check its length to find out if the user did set the `deletion_delay_hours` value.  Please let me know if you feel there is any other better approach to handle this use case.

[the remark this answers is not in the document]
Comment (melinath) [reply mmv1/products/vmwareengine/PrivateCloud.yaml:95]: We don't have a strong pattern for this kind of situation but we usually use a second boolean field to control behavior in some way. For example:

- `send_deletion_delay_hours_if_zero`: similar to https://github.com/GoogleCloudPlatform/magic-modules/pull/10717. It would mean that people who want to send zero need to really really mean it!
- you could also do a field like `use_default_delay_hours` to indicate that the server should by default always send whatever is set (including zero) unless the field is set.

I'm guessing you'd prefer something like the former since folks are less likely to set `0` for delay_hours accidentally.

[the remark this answers is not in the document]
Comment (swamitagupta) [reply mmv1/products/vmwareengine/PrivateCloud.yaml:95]: I am concerned that introducing an additional boolean field alongside the delay hours field could potentially introduce user error, particularly given its absence from the VMware Engine API.

Creating a nested object looks like a cleaner and more adaptive approach. This would also hold additional delete API parameters which might be added in the near future.

Wdyt?

File: mmv1/templates/terraform/post_delete/private_cloud.go.erb (hunk 1 above)
Comment (swamitagupta) [reply mmv1/templates/terraform/post_delete/private_cloud.go.erb:31]: Noted, thanks

File: mmv1/templates/terraform/pre_create/vmwareengine_private_cloud.go.erb [hunk 4]
```diff
@@ -0,0 +1,9 @@
+// Check if the project exists in a deleted state
+pcMarkedForDeletion, err := isPrivateCloudInDeletedState(config, d, billingProject, userAgent)
+if err != nil {
+	return fmt.Errorf("Error checking if Private Cloud exists and is marked for deletion: %s", err)
+}
+if pcMarkedForDeletion {
+    log.Printf("[DEBUG] Private Cloud exists and is marked for deletion. Triggering UNDELETE of the Private Cloud.\n")
+    url = url + ":undelete"
```
[the remark this answers is not in the document]
Comment (swamitagupta) [reply mmv1/templates/terraform/pre_create/vmwareengine_private_cloud.go.erb:8]: Noted, removed

File: mmv1/products/vmwareengine/PrivateCloud.yaml [hunk 5]
```diff
@@ -89,6 +91,17 @@ parameters:
     url_param_only: true
     description: |
       The ID of the PrivateCloud.
+  - !ruby/object:Api::Type::NestedObject
```
[the remark this answers is not in the document]
Comment (swamitagupta) [reply mmv1/products/vmwareengine/PrivateCloud.yaml:94]: Modified, thanks

File: mmv1/products/vmwareengine/PrivateCloud.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (swamitagupta) [reply mmv1/products/vmwareengine/PrivateCloud.yaml]: Used `send_deletion_delay_hours_if_zero` since virtual fields dont support nested fields at the moment

File: mmv1/products/vmwareengine/PrivateCloud.yaml [hunk 6]
```diff
@@ -90,6 +91,16 @@ parameters:
     description: |
       The ID of the PrivateCloud.
 
+virtual_fields:
+  - !ruby/object:Api::Type::Integer
+    name: "deletion_delay_hours"
+    description: |
+      The number of hours to delay this request. You can set this value to an hour between 0 to 8, where setting it to 0 starts the deletion request immediately. If no value is set, a default value is set at the API Level.
```
Comment (c2thorn) [inline mmv1/products/vmwareengine/PrivateCloud.yaml:98]: What is the API default value for deletion delay? Is it 0? How often do you think users will need to additionally set `send_deletion_delay_hours_if_zero`?

File: mmv1/templates/terraform/constants/vmwareengine_private_cloud.go.erb [hunk 7]
```diff
@@ -0,0 +1,33 @@
+func vmwareenginePrivateCloudStandardTypeDiffSuppressFunc(_, old, new string, d *schema.ResourceData) bool {
+	if (old == "STANDARD" && new == "") || (old == "" && new == "STANDARD") {
+		return true
+	}
+    if (isMultiNodePrivateCloud(d) && old == "TIME_LIMITED" && new == "STANDARD") {
+        log.Printf("[DEBUG] Multinode Private Cloud found, facilitating TYPE change to STANDARD")
```
Comment (c2thorn) [inline mmv1/templates/terraform/constants/vmwareengine_private_cloud.go.erb:6]: Can you give more context as to why a type diff from `TIME_LIMITED` -> `STANDARD` needs to be suppressed with Multinode? Also, is this related or separate to delayed deletion?

File: mmv1/third_party/terraform/services/vmwareengine/resource_vmwareengine_private_cloud_test.go [hunk 8]
```diff
@@ -18,7 +18,7 @@ func TestAccVmwareenginePrivateCloud_vmwareEnginePrivateCloudUpdate(t *testing.T
 	t.Parallel()
 
 	context := map[string]interface{}{
-		"region":          "southamerica-west1",
+		"region":          "me-west1",
```
Comment (c2thorn) [inline mmv1/third_party/terraform/services/vmwareengine/resource_vmwareengine_private_cloud_test.go:21]: can you provide more context as to why this is being updated?

File: mmv1/third_party/terraform/services/vmwareengine/resource_vmwareengine_subnet_test.go [hunk 9]
```diff
@@ -8,6 +8,8 @@ import (
 )
 
 func TestAccVmwareengineSubnet_vmwareEngineUserDefinedSubnetUpdate(t *testing.T) {
+	// Temporarily skipping so that this test does not run and consume resources during PR pushes. It is bound to fail and is being fixed by PR #10992
```
Comment (c2thorn) [inline mmv1/third_party/terraform/services/vmwareengine/resource_vmwareengine_subnet_test.go:11]: we will remove these before submitting right?

**Review by c2thorn — COMMENTED** [review]
asked some clarifying questions for context, things seem okay at first glance however.
Lots of parts, but since changes are contained to a diff suppress or deletion functionality, it should be fairly safe.

File: mmv1/products/vmwareengine/PrivateCloud.yaml (hunk 6 above)
Comment (swamitagupta) [reply mmv1/products/vmwareengine/PrivateCloud.yaml:98]: Currently, the default delay at the API level is set to 3 hours. However, it might be modified in the near future. 

Users would need to set `send_deletion_delay_hours_if_zero` if they want to send `delay_hours=0` through the API call and delete the PC immediately. I have added this boolean field after a discussion with Stephen in a previous thread since terraform does not differentiate between "no value set" and "0 value set".

File: mmv1/third_party/terraform/services/vmwareengine/resource_vmwareengine_private_cloud_test.go (hunk 8 above)
Comment (swamitagupta) [reply mmv1/third_party/terraform/services/vmwareengine/resource_vmwareengine_private_cloud_test.go:21]: Some months back, we exhausted the number of nodes in the southamerca-west1 region which were being used for this test, leading to PC creation failures in multiple terraform tests. 

As a result, [go/gcve-resource-optimisation-for-tf-tests](http://goto.google.com/gcve-resource-optimisation-for-tf-tests) was created and additional nodes were reserved in me-west1 region for terraform test projects. This PR is part of its implementation efforts along with PR #10992.

File: mmv1/third_party/terraform/services/vmwareengine/resource_vmwareengine_subnet_test.go (hunk 9 above)
Comment (swamitagupta) [reply mmv1/third_party/terraform/services/vmwareengine/resource_vmwareengine_subnet_test.go:11]: We can keep these commented till PR #10992 gets submitted. That way, the consumption of nodes for tests which are bound to fail would be prevented. 

These tests would be modified and combined by PR #10992 which is built on the design [go/gcve-resource-optimisation-for-tf-tests](http://goto.google.com/gcve-resource-optimisation-for-tf-tests)

File: mmv1/templates/terraform/constants/vmwareengine_private_cloud.go.erb (hunk 7 above)
Comment (c2thorn) [reply mmv1/templates/terraform/constants/vmwareengine_private_cloud.go.erb:6]: I also see that `type` is set to `ignore_read`, so at least it will remain consistent. Thanks for the context.

File: mmv1/third_party/terraform/services/vmwareengine/resource_vmwareengine_subnet_test.go (hunk 9 above)
Comment (c2thorn) [reply mmv1/third_party/terraform/services/vmwareengine/resource_vmwareengine_subnet_test.go:11]: Understood, o the relevant test for these code changes are all in `mmv1/third_party/terraform/services/vmwareengine/resource_vmwareengine_private_cloud_test.go`

File: mmv1/products/vmwareengine/PrivateCloud.yaml (hunk 6 above)
Comment (c2thorn) [reply mmv1/products/vmwareengine/PrivateCloud.yaml:98]: Just noting that I do not like the need for `send_deletion_delay_hours_if_zero` but cannot think of another way to force send 0 only when the user has specified due to TF limitations. I am sure this was discussed previously

**Review by c2thorn — APPROVED** [review]
LGTM

File: mmv1/products/vmwareengine/PrivateCloud.yaml (hunk 6 above)
Comment (swamitagupta) [reply mmv1/products/vmwareengine/PrivateCloud.yaml:98]: Yes, [this](https://github.com/GoogleCloudPlatform/magic-modules/pull/10764#discussion_r1655305161) is the previous discussion thread for implementing this change. 

Another approach which I could think of was using `delay_hours` inside a nested object `deletion_parameters` but nested objects are not yet supported for `virtual_fields`.

File: mmv1/third_party/terraform/services/vmwareengine/resource_vmwareengine_subnet_test.go (hunk 9 above)
Comment (swamitagupta) [reply mmv1/third_party/terraform/services/vmwareengine/resource_vmwareengine_subnet_test.go:11]: Yes, thats correct
